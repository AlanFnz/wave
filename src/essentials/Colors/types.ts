import { Object, String, List } from 'ts-toolbelt';

type HSL = `hsl(${number}, ${number}%, ${number}%)` | `hsla(${number}, ${number}%, ${number}%, ${number})`;
type RGB = `rgb(${number}, ${number}, ${number})` | `rgba(${number}, ${number}, ${number}, ${number})`;

export type Color = HSL | RGB;

export interface SemanticColorsSchema {
    text: {
        primary: Color,
        primaryInverted: Color,
        secondary: Color,
        secondaryInverted: Color,
        tertiary: Color,
        tertiaryInverted: Color,
        disabled: Color,
        disabledInverted: Color,
        link: Color,
        linkInverted: Color,
        linkHover: Color,
        linkHoverInverted: Color,
        info: Color,
        danger: Color,
        dangerInverted: Color,
        success: Color,
        warning: Color,
    },
    background: {
        primary: {
            default: Color,
            inverted: Color,
            emphasized: Color,
            hover: Color,
            disabled: Color,
        },   
        secondary: {
            default: Color,
            emphasized: Color,
        },
        danger: {
            default: Color,
            emphasized: Color,
            hover: Color,
        },
        info: {
            default: Color,
            emphasized: Color,
            hover: Color,
        },
        success: {
            default: Color,
            emphasized: Color,
        },
        warning: {
            default: Color,
            emphasized: Color,
        },
        transparent: 'transparent',
    },
    icon: {
        primary: {
            default: Color,
            inverted: Color,
        },
        secondary: {
            default: Color,
            inverted: Color,
            hover: Color,
            hoverInverted: Color,
        },
        tertiary: {
            default: Color,
            inverted: Color,
        },
        disabled: {
            default: Color,
            inverted: Color,
        },
        action: {
            default: Color,
            emphasized: Color,
        },
        danger: {
            default: Color,
            emphasized: Color,
        },
        success: Color,
        warning: Color,
        info: Color,
    },
    border: {
        primary: {
            default: Color,
            emphasized: Color,
            inverted: Color,
        },
        secondary: {
            default: Color,
            inverted: Color,
        },
        disabled: {
            default: Color,
            inverted: Color,
        },
        info: {
            default: Color,
            emphasized: Color,
        },
        success: {
            default: Color,
            emphasized: Color,
        },
        warning: {
            default: Color,
            emphasized: Color,
        },
        danger: {
            default: Color,
            emphasized: Color,
            disabled: Color,
        },
    },
}

export type SemanticCssToken = String.Join<List.Required<Object.Paths<SemanticColorsSchema>>, '-'>
export type SemanticCssVariable = `--wave-s-color-${SemanticCssToken}`;

export type ReadCssVariable = `var(${SemanticCssVariable})` | `var(${SemanticCssVariable}, ${string})`;

export type ComponentSemanticTokens = {
    [key: string]: ReadCssVariable | ComponentSemanticTokens;
}