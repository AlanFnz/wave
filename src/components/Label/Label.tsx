import { FC } from 'react';
import styled from 'styled-components';
import { compose, margin, MarginProps, ResponsiveValue, variant } from 'styled-system';
import { Text } from '../Text/Text';
import { Colors } from '../../essentials';
import { theme } from '../../essentials/theme';
import { get } from '../../utils/themeGet';

interface LabelProps extends MarginProps {
    /**
     * Set the appropriate colors for the component with 'info' as a default
     */
    variant?: ResponsiveValue<'info' | 'success' | 'danger'>;
}

const variantStyles = variant({
    variants: {
        info: {
            color: Colors.AUTHENTIC_BLUE_900,
            borderColor: Colors.AUTHENTIC_BLUE_200,
            backgroundColor: Colors.AUTHENTIC_BLUE_50
        },
        success: {
            color: Colors.POSITIVE_GREEN_1000,
            borderColor: Colors.POSITIVE_GREEN_350,
            backgroundColor: Colors.POSITIVE_GREEN_50
        },
        danger: {
            color: Colors.NEGATIVE_ORANGE_1000,
            borderColor: Colors.NEGATIVE_ORANGE_350,
            backgroundColor: Colors.NEGATIVE_ORANGE_50
        }
    }
});

const Label: FC<LabelProps> = styled(Text).attrs({ theme })<LabelProps>`
    display: inline-block;
    box-sizing: border-box;
    font-size: 0.625rem;
    border-radius: ${get('radii.1')};
    font-weight: ${get('fontWeights.semibold')};
    padding: 0.1875rem 0.5rem;

    color: ${Colors.AUTHENTIC_BLUE_900};
    border: 0.0625rem solid ${Colors.AUTHENTIC_BLUE_200};
    background-color: ${Colors.AUTHENTIC_BLUE_50};

    ${compose(variantStyles, margin)}
`;

export { Label, LabelProps };
