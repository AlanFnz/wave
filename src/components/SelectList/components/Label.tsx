import { ReactNode } from 'react';
import styled from 'styled-components';

import { getSemanticValue } from '../../../utils/cssVariables';
import { get } from '../../../utils/themeGet';
import { variantStyles, errorStyles, disabledStyles } from '../styles';
import { Variant, Size } from '../types';

interface LabelProps {
    inverted?: boolean;
    error?: boolean;
    isDisabled?: boolean;
    size: Size;
    variant: Variant;
    children: ReactNode;
}

export const Label = styled.label<LabelProps>`
    position: absolute;
    pointer-events: none;
    line-height: 1;
    font-weight: ${get('fontWeights.semibold')};
    background: ${p => getSemanticValue(p.inverted ? 'background-element-primary-default' : 'background-element-neutral-default')};
    color: ${p => getSemanticValue(p.inverted ? 'foreground-neutral-faded' : 'foreground-neutral-emphasized')};
    transition: color 125ms ease;

    ${p => p.error && errorStyles.label()}
    ${({ isDisabled }) => isDisabled && disabledStyles.label}
    ${({ variant, size }) => variantStyles.label({ variant, size })}
`;
