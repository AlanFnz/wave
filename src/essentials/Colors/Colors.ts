export enum Colors {
    WHITE = '#FFFFFF',
    AUTHENTIC_BLUE_1100 = '#000F1F',
    AUTHENTIC_BLUE_900 = '#001E3E',
    AUTHENTIC_BLUE_550 = '#637689',
    AUTHENTIC_BLUE_350 = '#9CA7B4',
    AUTHENTIC_BLUE_200 = '#C6CDD4',
    AUTHENTIC_BLUE_50 = '#F1F2F4',
    FREEDOM_RED_1000 = '#BF0820',
    FREEDOM_RED_900 = '#FF0A2B',
    ACTION_BLUE_1000 = '#0750A4',
    ACTION_BLUE_900 = '#096BDB',
    ACTION_BLUE_350 = '#9FC5F1',
    ACTION_BLUE_150 = '#D6E6F9',
    ACTION_BLUE_100 = '#E4EFFB',
    ACTION_BLUE_50 = '#F1F7FD',
    BUMPY_MAGENTA_1000 = '#BF00B9',
    BUMPY_MAGENTA_900 = '#FF00F7',
    BUMPY_MAGENTA_350 = '#FF9CFC',
    BUMPY_MAGENTA_50 = '#FFF1FF',
    POSITIVE_GREEN_1000 = '#05763B',
    POSITIVE_GREEN_900 = '#069D4F',
    POSITIVE_GREEN_350 = '#9ED9BB',
    POSITIVE_GREEN_50 = '#F1FAF5',
    ATTENTION_YELLOW_900 = '#FFCC00',
    ATTENTION_YELLOW_350 = '#FFEB9C',
    ATTENTION_YELLOW_50 = '#FFFCF1',
    NEGATIVE_ORANGE_1000 = '#BF4704',
    NEGATIVE_ORANGE_900 = '#FF5E05',
    NEGATIVE_ORANGE_350 = '#FFC09E',
    NEGATIVE_ORANGE_50 = '#FFF6F1'
}
