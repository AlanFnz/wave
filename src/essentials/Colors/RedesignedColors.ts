import { SemanticColorsSchema } from './types';

// Global Tier (--wave-g-color-...)
export const Colors = {
    white: 'hsla(0, 0%, 99%, 1)',
    black: 'hsla(0, 0%, 0%, 1)',
    blue: {
        primary: {
            1100: 'hsla(0, 0%, 0%, 1)',
            900: 'hsla(350, 10%, 11%, 1)',
            550: 'hsla(0, 2%, 38%, 1)',
            350: 'hsla(0, 1%, 55%, 1)',
            200: 'hsla(0, 1%, 73%, 1)',
            50: 'hsla(0, 2%, 91%, 1)',
        },
        secondary: {
            1000: 'hsla(346, 51%, 33%, 1)',
            900: 'hsla(347, 42%, 43%, 1)',
            350: 'hsla(349, 89%, 76%, 1)',
            150: 'hsla(351, 100%, 85%, 1)',
            100: 'hsla(352, 100%, 93%, 1)',
            50: 'hsla(9, 100%, 96%, 1)',
        }
    },
    red: {
        1000: 'hsla(350, 91%, 41%, 1)',
        900: 'hsla(347, 42%, 43%, 1)',
    },
    magenta: {
        1000: 'hsla(302, 100%, 18%, 1)',
        900: 'hsla(302, 100%, 33%, 1)',
        350: 'hsla(311, 100%, 84%, 1)',
        50: 'hsla(324, 100%, 96%, 1)',
    },
    green: {
        1000: 'hsla(146, 100%, 11%, 1)',
        900: 'hsla(149, 98%, 22%, 1)',
        350: 'hsla(134, 55%, 68%, 1)',
        50: 'hsla(128, 94%, 94%, 1)',
    },
    yellow: {
        900: 'hsla(48, 100%, 47%, 1)',
        350: 'hsla(42, 100%, 79%, 1)',
        50: 'hsla(30, 100%, 97%, 1)',
    },
    orange: {
        1000: 'hsla(356, 100%, 29%, 1)',
        900: 'hsla(0, 75%, 42%, 1)',
        350: 'hsla(6, 100%, 84%, 1)',
        50: 'hsla(9, 100%, 96%, 1)'
    },
    transparent: 'transparent',
} as const;

// Semantic Colors Tier (--wave-s-color-...)
export const SemanticColors = {
    text: {
        primary: Colors.blue.primary[900],
        primaryInverted: Colors.white,
        secondary: Colors.blue.primary[550],
        secondaryInverted: Colors.blue.primary[200],
        tertiary: Colors.blue.primary[350],
        tertiaryInverted: Colors.blue.primary[350],
        disabled: Colors.blue.primary[200],
        disabledInverted: Colors.blue.primary[550],
        link: Colors.blue.secondary[900],
        linkInverted: Colors.white,
        linkHover: Colors.blue.secondary[1000],
        linkHoverInverted: Colors.blue.secondary[50],
        danger: Colors.orange[1000],
        dangerInverted: Colors.orange[900],
        success: Colors.green[900],
        warning: Colors.yellow[900],
        info: Colors.blue.primary[550],
    },
    background: {
        primary: {
            default: Colors.white,
            inverted: Colors.blue.primary[900],
            emphasized: Colors.blue.primary[900],
            hover: Colors.blue.primary[1100],
            disabled: Colors.blue.primary[200],
        },
        secondary: {
            default: Colors.blue.primary[50],
            emphasized: Colors.blue.primary[550],
        },
        danger: {
            default: Colors.orange[50],
            emphasized: Colors.orange[900],
            hover: Colors.orange[1000]
        },
        info: {
            default: Colors.blue.primary[50],
            emphasized: Colors.blue.primary[550],
            hover: Colors.blue.secondary[350],
        },
        success: {
            default: Colors.green[50],
            emphasized: Colors.green[900],
        },
        warning: {
            default: Colors.yellow[50],
            emphasized: Colors.yellow[900],
        },
        transparent: Colors.transparent,
    },
    icon: {
        primary: {
            default: Colors.blue.primary[900],
            inverted:  Colors.white,
        },
        secondary: {
            default: Colors.blue.primary[350],
            inverted: Colors.blue.primary[50],
            hover: Colors.blue.primary[900],
            hoverInverted: Colors.white,
        },
        tertiary: {
            default: Colors.blue.primary[350],
            inverted: Colors.blue.primary[350],
        },
        disabled: {
            default: Colors.blue.primary[200],
            inverted: Colors.blue.primary[550],
        },
        action: {
            default: Colors.blue.secondary[900],
            emphasized: Colors.blue.secondary[1000],
        },
        danger: {
            default: Colors.orange[900],
            emphasized: Colors.orange[1000],
        },
        success: Colors.green[900],
        warning: Colors.yellow[900],
        info: Colors.blue.primary[350],
    },
    border: {
        primary: {
            default: Colors.blue.primary[200],
            emphasized: Colors.blue.primary[1100],
            inverted: Colors.blue.primary[550],
        },
        secondary: {
            default: Colors.blue.primary[50],
            inverted: Colors.white,
        },
        disabled: {
            default: Colors.blue.primary[200],
            inverted: Colors.blue.primary[550],
        },
        info: {
            default: Colors.blue.primary[200],
            emphasized: Colors.blue.primary[550],
        },
        success: {
            default: Colors.green[350],
            emphasized: Colors.green[900],
        },
        warning: {
            default: Colors.yellow[350],
            emphasized: Colors.yellow[900]
        },
        danger: {
            default: Colors.orange[350],
            emphasized: Colors.orange[900],
            disabled: Colors.orange[350],
        },
    },
} satisfies SemanticColorsSchema