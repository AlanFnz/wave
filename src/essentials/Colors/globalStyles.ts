// import { createGlobalStyle } from 'styled-components';
// import { SemanticColors as ClassicSemanticColors } from './Colors';
// import { SemanticColors as RedesignedSemanticColors } from './RedesignedColors';
// import { generateCssVariables } from '../../utils/generateCssVariables';

// export const ClassicColors = createGlobalStyle`
//   :root {
//     /* ${generateCssVariables(ClassicSemanticColors)} */
//   }
// `;

// export const RedesignedColors = createGlobalStyle`
//   :root {
//     /* ${generateCssVariables(RedesignedSemanticColors)} */
//   }
// `;
