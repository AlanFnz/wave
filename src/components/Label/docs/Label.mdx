---
name: Label
menu: Components
route: /components/label
---

import { Playground } from 'docz';
import { Label } from '../Label.tsx';
import { LabelPropsTable } from './LabelPropsTable';
import { Combination } from '../../../docs/Combination';
import { StyledSystemLinks } from '../../../docs/StyledSystemLinks';

# Label

Multiple or single labels can be used to categorize items.
Use short labels for easy scanning. Use two words only if necessary to describe the status and differentiate it from other tags.

## Properties

<LabelPropsTable />
<br/>
<StyledSystemLinks component="Label" supportedProps={ ['margin'] }/>


## Examples

<Playground>
    <Label variant="success">Label</Label>
</Playground>

## Combinations

<Combination variant={["info", "success", "danger"]}>
    {(props, i) => <Label key={i} {...props}>Label</Label>}
</Combination>

