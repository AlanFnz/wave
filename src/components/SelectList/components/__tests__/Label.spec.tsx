import React from 'react';
import { render } from '@testing-library/react';

import { Label } from '../Label';
import { Colors } from '../../../../essentials/Colors/Colors';

describe('Label', () => {
    it('renders small boxed label', () => {
        const { container } = render(
            <Label size="small" variant="boxed">
                Label
            </Label>
        );

        expect(container).toMatchSnapshot();
    });

    it('renders medium boxed label', () => {
        const { container } = render(
            <Label size="medium" variant="boxed">
                Label
            </Label>
        );

        expect(container).toMatchSnapshot();
    });

    it('renders small bottom-lined label', () => {
        const { container } = render(
            <Label size="small" variant="bottom-lined">
                Label
            </Label>
        );

        expect(container).toMatchSnapshot();
    });

    it('renders medium bottom-lined label', () => {
        const { container } = render(
            <Label size="medium" variant="bottom-lined">
                Label
            </Label>
        );

        expect(container).toMatchSnapshot();
    });

    it('renders error label', () => {
        const { container } = render(
            <Label size="medium" variant="boxed" error>
                Label
            </Label>
        );

        expect(container.firstChild).toHaveStyleRule('color', Colors.NEGATIVE_ORANGE_900);
    });

    it('renders disabled label', () => {
        const { container } = render(
            <Label size="medium" variant="boxed" isDisabled>
                Label
            </Label>
        );

        expect(container.firstChild).toHaveStyleRule('color', Colors.AUTHENTIC_BLUE_200);
    });

    it('renders disabled inverted label', () => {
        const { container } = render(
            <Label size="medium" variant="boxed" inverted isDisabled>
                Label
            </Label>
        );

        expect(container.firstChild).toHaveStyleRule('color', Colors.AUTHENTIC_BLUE_550);
    });

    it('renders inverted label', () => {
        const { container } = render(
            <Label size="medium" variant="boxed" inverted>
                Label
            </Label>
        );

        expect(container.firstChild).toHaveStyleRule('color', Colors.AUTHENTIC_BLUE_200);

        expect(container.firstChild).toHaveStyleRule('background', Colors.AUTHENTIC_BLUE_900);
    });
});
