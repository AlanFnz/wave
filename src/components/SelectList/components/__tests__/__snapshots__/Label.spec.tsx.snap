// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Label renders medium bottom-lined label 1`] = `
.c0 {
  position: absolute;
  pointer-events: none;
  line-height: 1;
  font-weight: 600;
  background: #FFFFFF;
  color: #637689;
  -webkit-transition: color 125ms ease;
  transition: color 125ms ease;
  top: 0.375rem;
  left: 0;
  padding: 0 0.25rem;
  font-size: 0.75rem;
}

<div>
  <label
    class="c0"
  >
    Label
  </label>
</div>
`;

exports[`Label renders medium boxed label 1`] = `
.c0 {
  position: absolute;
  pointer-events: none;
  line-height: 1;
  font-weight: 600;
  background: #FFFFFF;
  color: #637689;
  -webkit-transition: color 125ms ease;
  transition: color 125ms ease;
  top: -0.3125rem;
  left: 0.5625rem;
  padding: 0 0.25rem;
  font-size: 0.75rem;
}

<div>
  <label
    class="c0"
  >
    Label
  </label>
</div>
`;

exports[`Label renders small bottom-lined label 1`] = `
.c0 {
  position: absolute;
  pointer-events: none;
  line-height: 1;
  font-weight: 600;
  background: #FFFFFF;
  color: #637689;
  -webkit-transition: color 125ms ease;
  transition: color 125ms ease;
  top: 0.125rem;
  left: 0;
  padding: 0 0.125rem;
  font-size: 0.625rem;
}

<div>
  <label
    class="c0"
  >
    Label
  </label>
</div>
`;

exports[`Label renders small boxed label 1`] = `
.c0 {
  position: absolute;
  pointer-events: none;
  line-height: 1;
  font-weight: 600;
  background: #FFFFFF;
  color: #637689;
  -webkit-transition: color 125ms ease;
  transition: color 125ms ease;
  top: -0.3125rem;
  left: 0.4375rem;
  padding: 0 0.125rem;
  font-size: 0.625rem;
}

<div>
  <label
    class="c0"
  >
    Label
  </label>
</div>
`;
